New request handler

erickvermot opened the pull request:
refers to #292.
depends on PR https://github.com/kytos/kytos-napps/pull/217

erickvermot committed:
New request handler
implements a request handler that does not depend on the python-openflow lib, leaving message processing to of_core.

erickvermot commented:
TODO: move constants to settings

diraol commented on kytos/core/switch.py:
```diff
@@ -191,6 +192,24 @@ def as_json(self):
         return json.dumps(self.as_dict())
 
 
+class Connection_states(Enum):
```
`s/Connection_states/CONNECTION_STATE/`

Despite being a class, it will behave like a "constant", therefore I think we can use constants naming rules. @cemsbr Am I right here?
Or should we use `ConnectionState` instead?

erickvermot replied:
I agree with the caps, I wrote with caps first but the changed it since it was a class...
I'll capitalize it back

diraol commented on kytos/core/switch.py:
```diff
@@ -191,6 +192,24 @@ def as_json(self):
+    FINISHED = 3
+
+
+class Protocol_info():
```
s/Protocol_info/Protocol/

(I was rewriting it to `ProtocolInfo`, but it is very difficult to distinguish the lowercase `L` and the uppercase `i`, so I think I just prefer `Protocol`. Or, eventually, `ProtocolData`, despite the fact that this is not a very good semantic name for its purpose.)

erickvermot replied:
I think just ``Protocol`` is ok, specially if we add links to pack and unpack functions later.

diraol commented on kytos/core/switch.py:
```diff
@@ -207,6 +226,23 @@ def __init__(self, address, port, socket, switch=None):
         self.socket = socket
         self.switch = switch
+        self._state = None
+        self.STATES = Connection_states
```
I don't like the definition of a `CONSTANT` inside the class as an attribute.

I really prefer using the `CONNECTION_STATE` class directly when needed. And also, I think that makes sense to define the default value from `state` as `CONNECTION_STATE.NEW` value.

Regarding the `@property` and the `setter`, we are following a "pythonic" approach on the project, which means no fixed types (as python) and no excessive checks like C++/Java would require. The user are responsible for what it is doing.
Being so, I think that we can just have the `state` attribute (and document it in a way that our users knows that this must be one option from `CONNECTION_STATE`).

diraol replied:
But I am not absolutely sure about not using the `setter` here, just because of the "changed state logging". @beraldoleal and @cemsbr what do you think about it?

erickvermot replied:
I don't see it as a definition, but simply as a shortcut so that when we receive the event as an arg on a handler, we can access the possible states without having to import something new... I'll remove it if you guys want.

About the setter, I thought it was a good way to provide the logging.

diraol replied:
If we are going to use the setter here, the best approach would be something like this:
```
class Test():
    def __init__(self):
        self.x = None
    @property
    def x(self):
        return self.__x
    @x.setter
    def x(self, x):
        self.__x = x
        print('setting x with {}'.format(x))
```

diraol commented on kytos/core/tcp_server.py:
```diff
@@ -44,113 +40,108 @@ def __init__(self, server_address, RequestHandlerClass,
+        SOCKET_BIND_N_TRIES = 2
+        # Tries to bind the socket n times, it may still be
+        # in WAIT state from previous connection.
+        for i in range(SOCKET_BIND_N_TRIES):
```
@erickvermot take a look at the end of the section https://docs.python.org/3/library/socket.html#example. There it talks about `TIME_WAIT` on sockets, and the use of `socket.SO_REUSEADDR`. This may fix the problem you were trying to address here with this "loop" and it seems to be a more appropriate solution to the problem than an arbitrary loop.

erickvermot replied:
The example in the library forces the reuse of the address without waiting for the TIME_WAIT to expire. I think it would be better to wait for it before binding again.

diraol replied:
Hum.... I still not sure if what way to go.
@beraldoleal ? @cemsbr ? @renanrodrigo ? @abaruchi ?
Any thoughts?
Just adding one more link:
http://www.unixguide.net/network/socketfaq/4.5.shtml

Just to clarify for those that are just getting into the discussion.

After a socket is closed, it's state is changed to "closed" but it remains opened for some time (`TIME_WAIT`). This is a default behavior from Unix sockets.
If anyone tries to establish a new socket on the same "socket"*, then the OS will raise `Address already in use` error.

We have three options here.
1) Just raise the error;
2) Use a `timeout` and keep trying to bind to the socket again. [ @erickvermot proposal on this PR ];
3) Use the `socket.SO_REUSEADDR` to tell the OS that it should get the living (but closed) socket and "restart" it.

*A socket is a 5 tuple (`proto`, `local addr`, `local port`, `remote addr`, `remote port`)

renanrodrigo replied:
I agree with @erickvermot on option 2.
`socket.SO_REUSEADDR` could cause problems in case of quick switch reconnections if they are exchanging data with the controller.

diraol replied:
We ( @erickvermot , @abaruchi , @renanrodrigo and I) discussed this and the possibility of errors due to reusing the connection, considering the OpenFlow protocol (or eventually any other SDN protocol) is negligible. Even if it occurs, the consequences are minimal and manageable, in a way that the costs of waiting the `TIME_WAIT` to expire seems to be much higher than just reusing the socket and eventually dropping some lost packages or restarting the connection to a switch.

So, let's stick with the option 3 and reuse the socket without waiting the `TIME_WAIT`.

erickvermot replied:
I still believe there should be at least an option to deal with that like 'FORCEBIND' or similar. I'm not sure the consequences are negligible, and the default behaviour of TCP servers is to wait.
Since the others prefer to go with 3 and ignore the possible problems, I'll change the PR.

erickvermot replied:
@abaruchi @renanrodrigo @diraol @beraldoleal,
I've moved my solution to another branch in case someone wants to consider it: https://github.com/erickvermot/kytos/tree/socket_bind_fix_wait
It retries N times waiting X seconds between retries, and only forces the bind if the retries fails.
This gives at least some time (N*X) before forcing the bind, and may help prevent packets from an old connection and possibly from another switch to eventually be received by the new connection.

diraol commented on kytos/core/tcp_server.py:
```diff
@@ -44,113 +40,108 @@ def __init__(self, server_address, RequestHandlerClass,
-        This method builds a new connection, makes a new KytosEvent named
-        'kytos/core.connection.new' and put this in a raw buffer.
+        This method builds a new controller Connection, and places a
+        'kytos/core.connection.new' KytosEvent in the raw buffer.
```
`app` buffer

diraol commented on kytos/core/controller.py:
```diff
@@ -355,10 +355,14 @@ def msg_out_event_handler(self):
             destination = triggered_event.destination
-            if destination:
+            if (destination and
+                    not destination.state == destination.STATES.FINISHED):
```
`destination.state == CONNECTION_STATE.FINISHED`

diraol commented on kytos/core/controller.py:
```diff
@@ -355,10 +355,14 @@ def msg_out_event_handler(self):
                 self.log.debug("MsgOutEvent handler called")
+            else:
+                self.log.debug(
+                    "connection is finished. Cannot send message")
```
`finished` or `closed`? What is the best "networking word" for this @beraldoleal

erickvermot replied:
It was meant to indicate the 'FINISHED' state, which does not necessarily mean the socket is actually closed.

beraldoleal replied:
Let's use "Connection closed" because is an useful info to the network administrator. If we need more details like "FYN_WAIT" or something use as debug message.

diraol commented on kytos/core/tcp_server.py:
```diff
@@ -1,38 +1,34 @@
+    def __init__(self, server_address, RequestHandlerClass, controller):
+        """Constructor of KytosServer.
 
         Parameters:
```
@erickvermot as you are editing this file, could you please replace `Parameters` with `Args`? Tks!

diraol commented on kytos/core/tcp_server.py:
```diff
@@ -44,113 +40,108 @@ def __init__(self, server_address, RequestHandlerClass,
         self.port = self.client_address[1]
 
         self.connection = Connection(self.ip, self.port, self.request)
+        self.connection.state = self.connection.STATES.NEW
```
`self.connection.state = CONNECTION_STATE.NEW`

diraol commented on kytos/core/tcp_server.py:
```diff
@@ -44,113 +40,108 @@ def __init__(self, server_address, RequestHandlerClass,
-        log.info("Client %s:%s disconnected. Reason: %s",
+        log.info("Connection lost with Client %s:%s. Reason: %s",
                  self.ip, self.port, self.exception)
+        self.connection.state = self.connection.STATES.FINISHED
```
`self.connection.state = CONNECTION_STATE.FINISHED`

diraol commented on kytos/core/tcp_server.py:
```diff
@@ -44,113 +40,108 @@ def __init__(self, server_address, RequestHandlerClass,
+        log.info("Connection lost with Client %s:%s. Reason: %s",
                  self.ip, self.port, self.exception)
+        self.connection.state = self.connection.STATES.FINISHED
+        self.request.shutdown(SHUT_WR)
```
@erickvermot what about the `graceful shutdown` method we discussed?

erickvermot replied:
I'm implementing it, but first I want to be sure where it makes sense to call the shutdown and where it makes sense to call close directly.

diraol reviewed (commented):
There is still the need to set the `socket.SO_REUSEADDR` flag on the socket. Have you done it? I haven't seen it...

erickvermot commented:
well, seen or not, it's there...

beraldoleal commented on kytos/core/switch.py:
```diff
@@ -191,6 +193,24 @@ def as_json(self):
+    FINISHED = 3
+
+
+class Protocol():
```
`class Protocol:`

beraldoleal replied:
s/Protocol/ApplicationProtocol/g

beraldoleal commented on kytos/core/tcp_server.py:
```diff
@@ -44,113 +38,93 @@ def __init__(self, server_address, RequestHandlerClass,
         self.port = self.client_address[1]
 
         self.connection = Connection(self.ip, self.port, self.request)
+        self.connection.state = CONNECTION_STATE.NEW
```
Make `CONNECTION_STATE.NEW` as default. If yes, please remove this line.

beraldoleal commented on kytos/core/tcp_server.py:
```diff
@@ -44,113 +38,93 @@ def __init__(self, server_address, RequestHandlerClass,
         self.request.settimeout(30)
         self.exception = None
 
         event = KytosEvent(name='kytos/core.connection.new',
```
Please, emit the event according to the port. Ex: If port is 6633 the event should be `kytos/core.connection.new.openflow`

beraldoleal commented on kytos/core/tcp_server.py:
```diff
@@ -44,113 +38,93 @@ def __init__(self, server_address, RequestHandlerClass,
-            except InterruptedError as exception:
+                new_data = self.request.recv(MAX_SIZE)
+            except (SocketError, OSError, InterruptedError) as exception:
                 self.exception = exception
```
s/exception/last_exception/g

erickvermot replied:
It only catches one exception, should I change it all the same?

beraldoleal commented on kytos/core/tcp_server.py:
```diff
@@ -44,113 +38,93 @@ def __init__(self, server_address, RequestHandlerClass,
+                          exception)
+                break
+            if new_data == b'':
+                self.exception = 'Null Request (closed by client)'
```
"Request closed by client"

beraldoleal commented on kytos/core/tcp_server.py:
```diff
@@ -44,113 +38,93 @@ def __init__(self, server_address, RequestHandlerClass,
-
-            event = KytosEvent(name='kytos/core.messages.openflow.new',
+                       'new_data': new_data}
+            event = KytosEvent(name='kytos/core.server.data.in',
```
Here the same, add the port number on the event.

beraldoleal replied:
`kytos/core.(protocol|port).data.in`

beraldoleal commented on kytos/core/tcp_server.py:
```diff
@@ -44,113 +38,93 @@ def __init__(self, server_address, RequestHandlerClass,
+        log.info("Connection lost with Client %s:%s. Reason: %s",
                  self.ip, self.port, self.exception)
+        self.connection.state = CONNECTION_STATE.FINISHED
+        self.request.shutdown(SHUT_RDWR)
```
Please do a double check if is really necessary close after shutdown.

erickvermot commented:
I have updated the branch, but I'm experiencing some problems when I test it with mininet.
I do not have any problems when testing with a generic fake switch I made, but mininet seems to be sending many 'packet-in' to the controller, even though I'm not telling him to ping or anything.
I'm still investigating what is happening. 

\+ test is complaining that:
kytos/core/tcp_server.py:1:1: E0203 Access to member 'connection' before its definition line 82 [pylint]
but I believe the 'access' it it worrying about is the one in switch.py line 426 where we check if the connection exists...

diraol commented on kytos/core/switch.py:
```diff
@@ -207,6 +229,22 @@ def __init__(self, address, port, socket, switch=None):
+            raise Exception('Unknown State', new_state)
+        self._state = new_state # noqa
+        log.info('Connection %s changed state: %s',
+                  self.id, self.state)
```
There is an indentation error here (1 extra space on line 247)

diraol commented on kytos/core/tcp_server.py:
```diff
@@ -44,118 +37,112 @@ def __init__(self, server_address, RequestHandlerClass,
 
+        log.info("New connection from %s:%s", self.ip, self.port)
+
         self.connection = Connection(self.ip, self.port, self.request)
```
Add a `# noqa` on the end of this line to avoid `pylint` error `E0203 Access to member 'connection' before its definition`.

erickvermot replied:
adding #noqa on line 82 did not solve it, since tox complains on tcp_server line 1 (probably when importing..)

diraol replied:
O_o That's really strange..... I'll run it locally to see what is going on....

erickvermot commented:
@beraldoleal ping

beraldoleal commented:
@erickvermot the problems with mininet that you said also happen without this PR ?

Also you have to fix the pylint errors.

erickvermot commented:
@beraldoleal, about mininet, it was no problem at all, just lldp doing its job. But yes, it happens with or without this PR.

About the pylint errors, I talked about this a few weeks ago with @diraol, https://github.com/kytos/kytos/pull/374#issuecomment-298771159 and he also doesn't know how to fix it or how to make it be ignored... Any suggestion?
kytos/core/tcp_server.py:1:1: E0203 Access to member 'connection' before its definition line 82 [pylint]

About the commits, would you like me to combine the commits to a single one to make it more clean?

beraldoleal commented on kytos/core/tcp_server.py:
```diff
@@ -1,40 +1,33 @@
-    (defaults to 6633) for any package that comes from the network and then
-    stabilishes the socket connection to the devices (switches) when needed,
-    considering a thread per connection and keeping the connection alive.
+    The KytosServer will listen on the specified port (defaults to 6633)
```
Please, remove references to 6633 on this comments.

beraldoleal commented on kytos/core/switch.py:
```diff
@@ -191,6 +194,25 @@ def as_json(self):
+class ProtocolInfo:
+    """Class to hold simple protocol information for the connection."""
+
+    def __init__(self):
```
```
def __init__(self, name = None, version = None, state = None):
   self.name = name,
   ...
```

beraldoleal commented on kytos/core/switch.py:
```diff
@@ -191,6 +194,25 @@ def as_json(self):
+    FINISHED = 4
+
+
+class ProtocolInfo:
```
s/ProtocolInfo/ConnectionProtocol/g

beraldoleal commented on kytos/core/switch.py:
```diff
@@ -191,6 +194,25 @@ def as_json(self):
         return json.dumps(self.as_dict())
 
 
+class CONNECTION_STATE(Enum):
```
move `CONNECTION_STATE`, `ConnectionProtocol` and `Connection` to a new module , named `connection`

beraldoleal reviewed (changes requested):
One more time, please. This is the last one. :)

erickvermot committed:
minor changes discussed in meeting:
 changes related to nomenclature and moving of connection related classes to a new module named connection.py

erickvermot committed:
correct bad file descriptor errno
